### ISSUE TITLE
Pyright only seems to look at last variant in union of generic `dataclass`es

### ISSUE THREAD

**Issue by colehaus** [microsoft/pyright#5201]
With code like this:

```py
from dataclasses import dataclass
from typing import Generic, TypeVar

A = TypeVar("A")
@dataclass()
class DcWrapper(Generic[A]):
    value: A

test1: DcWrapper[int] | DcWrapper[str] = DcWrapper(1) # Errors, unexpectedly
test2: DcWrapper[int] | DcWrapper[str] = DcWrapper("ab")
test3: DcWrapper[int] | DcWrapper[str] = DcWrapper(None) # Errors, as expected

class Wrapper(Generic[A]):
    def __init__(self, value: A) -> None:
        self.value = value

test4: Wrapper[int] | Wrapper[str] = Wrapper(1)
test5: Wrapper[int] | Wrapper[str] = Wrapper("ab")
test6: Wrapper[int] | Wrapper[str] = Wrapper(None) # Errors, as expected
```
pyright 1.1.310 reports:
```
test.py:9:45 - error: Argument of type "Literal[1]" cannot be assigned to parameter "value" of type "str" in function "__init__"
    "Literal[1]" is incompatible with "str" (reportGeneralTypeIssues)
```

I would have expected `test1` to type check correctly and indeed mypy allows it.

Removing the `| Wrapper[str]` on the problematic line allows the line to type check successfully. And, as the code shows, this seems to work fine with non-`dataclass` `Generic`s. So it seems there is some problematic interaction here.

**Comment by colehaus** [comment]
```py
IntWrapper = NewType("IntWrapper", DcWrapper[int])
StrWrapper = NewType("StrWrapper", DcWrapper[str])

test1: DcWrapper[int] | DcWrapper[str] = IntWrapper(DcWrapper(1))
```
seems to work as a workaround.

**Comment by erictraut** [comment]
Thanks for the bug report. Although it is not immediately obvious, this issue is caused by the same underlying problem as https://github.com/microsoft/pyright/issues/5176 (another open bug).

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.311, which I just published. It will also be included in a future release of pylance.
